lib/ansible/galaxy/dependency_resolution/providers.py

```diff
@@ -339,15 +292,45 @@ def _find_matches(self, requirements):
                 for version, _none_src_server in coll_versions
             ]
 
+        def requirement_is_pinned(requirement):
+            return not (
+                requirement.ver == '*' or
+                requirement.ver.startswith('<') or
+                requirement.ver.startswith('>') or
+                requirement.ver.startswith('!=')
+            ) or requirement.ver[0].isdigit()
+
         latest_matches = []
         signatures = []
         extra_signature_sources = []  # type: list[str]
```

Comment:
_TODO:_ Figure out why this isn't reset on each outer loop cycle below.

Reply:
@s-hertel any ideas?

Reply:
@s-hertel this seems like a bug but I chose not to attempt addressing it within this PR. Still, it'd be good to figure it out...